### ISSUE TITLE
Flag use of Optional[FooType] without performing a None-check

### ISSUE THREAD

**Issue by davidfstr** [python/mypy#2176]
Given the following program:

```
from typing import Optional

maybe_str = None  # type: Optional[str]
print(len(maybe_str))
```

I would expect mypy to complain about the expression `len(maybe_str)` because len() is not a valid operation for the `None` case of a value of type `Optional[str]`, but it does not.

I have quite a lot of functions that return `Optional[FooType]`. When calling these functions I've frequently seen bugs in callers that don't do a None-check and try to blindly use the result. It would be valuable to me if such unchecked usages were flagged.

In cases where the caller asserted that the None case was impossible, the caller could use an `assert result not None` or `cast(FooType, result)` to explicitly signal to the typechecker that the None case is not expected.

Is this behavior intentional?

**Comment by gvanrossum** [comment]
Did you run mypy with `--strict-optional`? If not, mypy's (non-compliant) default behavior allows `None` everywhere. (Eventually `--strict-optional` will be the default, but it isn't yet.)

**Comment by davidfstr** [comment]
I was not aware of the `--strict-optional` argument. I will try it out. Thanks for the tip!

(Indeed my example program above is flagged when using the `--strict-optional` argument.)
